src/Storages/Statistics/Statistics.cpp

```diff
@@ -15,39 +18,134 @@ namespace ErrorCodes
 {
     extern const int LOGICAL_ERROR;
     extern const int INCORRECT_QUERY;
-    extern const int ILLEGAL_STATISTIC;
 }
 
-void MergeTreeStatisticsFactory::registerCreator(StatisticType stat_type, Creator creator)
+enum StatisticsFileVersion : UInt16
+{
+    V0 = 0,
+};
+
+/// Version / bitmask of statistics / data of statistics /
+
+ColumnStatistics::ColumnStatistics(const ColumnStatisticsDescription & stats_desc_)
+    : stats_desc(stats_desc_), rows(0)
 {
-    if (!creators.emplace(stat_type, std::move(creator)).second)
-        throw Exception(ErrorCodes::LOGICAL_ERROR, "MergeTreeStatisticsFactory: the statistic creator type {} is not unique", stat_type);
 }
 
-void MergeTreeStatisticsFactory::registerValidator(StatisticType stat_type, Validator validator)
+void ColumnStatistics::update(const ColumnPtr & column)
 {
-    if (!validators.emplace(stat_type, std::move(validator)).second)
-        throw Exception(ErrorCodes::LOGICAL_ERROR, "MergeTreeStatisticsFactory: the statistic validator type {} is not unique", stat_type);
+    rows += column->size();
+    for (const auto & iter : stats)
+    {
+        iter.second->update(column);
+    }
+}
 
+Float64 ColumnStatistics::estimateLess(Float64 val) const
+{
+    if (stats.contains(StatisticsType::TDigest))
+        return std::static_pointer_cast<TDigestStatistics>(stats.at(StatisticsType::TDigest))->estimateLess(val);
+    return rows * ConditionSelectivityEstimator::default_normal_cond_factor;
 }
 
-StatisticPtr TDigestCreator(const StatisticDescription & stat)
+Float64 ColumnStatistics::estimateGreater(Float64 val) const
 {
-    return StatisticPtr(new TDigestStatistic(stat));
+    return rows - estimateLess(val);
 }
 
-void TDigestValidator(const StatisticDescription &, DataTypePtr data_type)
+Float64 ColumnStatistics::estimateEqual(Float64 val) const
 {
-    data_type = removeNullable(data_type);
-    if (!data_type->isValueRepresentedByNumber())
-        throw Exception(ErrorCodes::ILLEGAL_STATISTIC, "TDigest does not support type {}", data_type->getName());
+    if (stats.contains(StatisticsType::Uniq) && stats.contains(StatisticsType::TDigest))
+    {
+        auto uniq_static = std::static_pointer_cast<UniqStatistics>(stats.at(StatisticsType::Uniq));
+        if (uniq_static->getCardinality() < 2048)
```

Comment:
The threshold seems arbitrary. Lets put at least a comment in so that future readers will not wonder what it is good for (e.g. I don't really understand it).